shynet/webserver.sh

```diff
@@ -1,8 +1,51 @@
 #!/bin/bash
 
+# Check if Setup is necessary, do setup as needed
+sanity_results=( $(python manage.py sanity_checks) )
+if [[ ${sanity_results[0]} == True ]]; then
+  echo "Running Migrations..."
+  {
+    python manage.py migrate && echo "Migrations Done"
+  } || {
+    echo "Failed Migrations, exiting" && exit 1
+  }
+  else
+  echo "Migrations Unecessary, skipping"
+fi
+if [[ ${sanity_results[1]} == True ]]; then
+  echo "Running CreateAdmin..."
+  {
+    temppwd=$( python manage.py registeradmin $SHYNET_EMAIL ) && echo "Admin Created, password $temppwd"
+  } || {
+    echo "Failed CreateAdmin, exiting" & exit 1
+  }
+  else
+  echo "CreateAdmin Unecessary, skipping"
+fi
+if [[ ${sanity_results[2]} == True ]]; then
+  echo "Setting Hostname..."
+  {
+    python manage.py hostname $SHYNET_HOST && echo "Host Set"
+  } || {
+    echo "Failed setting Hostname, exiting" & exit 1
+  }
+  else
+  echo "Hostname Unecessary, skipping"
+fi
+if [[ ${sanity_results[3]} == True ]]; then
+  echo "Setting Hostname..."
+  {
+    python manage.py whitelabel $SHYNET_NAME && echo "WhiteLabel Set"
+  } || {
+    echo "Failed Migrations, exiting" & exit 1
+  }
+  else
+  echo "WhiteLabel Unecessary, skipping"
+fi
+
```

Comment:
Can we move this into its own file? I'm concerned about putting this in `webserver.sh`. For multi-tenant systems, running migrations unexpectedly could have serious side-consequences.

Reply:
We can, no issue, but technically it only does the migrations if the tables don't exist, not if migrations are needed. If migrations are needed but the tables exist, the admin will have to run the commands manually, which I think is normal as per your comment.

As an aside the only way to have this run on first run is to have it be in the docker entrypoint. A resolution to this would therefore be to have another .sh "entrypoint.sh" which calls the checks and then calls webserver.sh.

Otherwise we'd need a less lightweight solution which would be to actually set a flag in postgres on first run that says "first run done", so that the checks don't ever get evaluated again after init, but that sounds like overkill to me... Thoughts ?

Reply:
I like the idea of having two entry points. Does docker-compose allow specifying container commands? If so, we can leave the unmodified webserver.sh as the default CMD, and add another script that the compose file can use which would run these checks and then the webserver script. Does that work?

Reply:
compose technically supports overriding the entrypoint, but only in the context of `run`, which is a one-time command.
`up` doesn't support entrypoint override, and is the standard way of running services with Compose.

The main "problem" here is that the shynet container itself is purely throwaway - it doesn't need to store anything persistent, that's what the postgre db is here for. So we have no volume to store the "yeah I ran once, don't run checks again and just call webserver.sh plz kthxbye" flag.
Now of course we could just create a volume for that, but that seems a bit... cumbersome, or store that flag in postgre itself, though that'll require a database callout to store the information of where the setup is (in case it gets interrupted).

The simplest way IMO is simply to:

- take migrations and spin it into its own migcheck.sh
- take the other checks and do "firstrun.sh" with it
- have an entrypoint.sh that:
	- runs firstrun.sh, and if any check comes back FALSE (meaning first run was done at least once), just calls webserver.sh
    - runs migcheck.sh only if all three checks before returned TRUE (meaning there's literally no admin user no nothing, so we can safely setup the database because it's empty)
    - uns webserver.sh as if both checks above see all's fine.

Reply:
... or we expose an environment variable FIRST_RUN and tie the checks to that being set to true manually in the Compose file. I'm actually OK with that on a design perspective - it'll work even if you want to spawn multiple instances of shynet and be easily adaptable to swarm mode.

Reply:
Another option is to change the default entry point to some kind of “launch” script that does these checks, and keep “webserver.css” exposed for those who want the server to run without side effects? I’m not opposed to changing the default script, as long as we keep “webserver.sh” available.

Reply:
that's what I meant with entrypoint.sh, lemme whip that up for ya, see if it matches your expectations more :)

Reply:
Ah got it, sounds good!